packages/backend-common/src/tokens/ServerTokenManager.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { JWKS, JWK, JWT } from 'jose';
+import { Config } from '@backstage/config';
+import { AuthenticationError } from '@backstage/errors';
+import { TokenManager } from './types';
+
+/**
+ * Creates and validates tokens for use during backend-to-backend
+ * authentication.
+ *
+ * @public
+ */
+export class ServerTokenManager implements TokenManager {
+  private readonly verificationKeys: JWKS.KeyStore;
+  private readonly signingKey: JWK.Key;
+
+  static noop() {
+    return new ServerTokenManager.NoopTokenManager();
+  }
+
+  static fromConfig(config: Config) {
+    return new ServerTokenManager(
+      config
+        .getConfigArray('backend.auth.keys')
+        .map(key => key.getString('secret')),
+    );
+  }
+
+  private constructor(secrets?: string[]) {
+    if (!secrets?.length) {
+      throw new Error(
+        'No secrets provided when constructing ServerTokenManager',
+      );
+    }
+
+    this.verificationKeys = new JWKS.KeyStore(
+      secrets.map(k => JWK.asKey({ kty: 'oct', k })),
+    );
+    this.signingKey = this.verificationKeys.all()[0];
+  }
+
+  async getToken(): Promise<{ token: string }> {
+    const jwt = JWT.sign({ sub: 'backstage-server' }, this.signingKey, {
+      algorithm: 'HS256',
+    });
+
+    return { token: jwt };
+  }
+
+  async authenticate(token: string): Promise<void> {
+    try {
+      JWT.verify(token, this.verificationKeys);
+    } catch (e) {
+      throw new AuthenticationError('Invalid server token');
+    }
+  }
+}
+
+export namespace ServerTokenManager {
+  export class NoopTokenManager implements TokenManager {
+    async getToken() {
+      return { token: '' };
```

Comment:
One limitation of the NoopTokenManager returning empty strings is that it makes it impossible to differentiate between unauthenticated requests from the backend and unauthenticated requests from elsewhere. This means that parts of the system that depend on backend-to-backend auth and need to branch based on whether a request comes from a user or the backend can only do so when a secret is configured.

Potentially the upshot here is that the permission framework docs will need to make configuring a secret a required part of setting it up, since without that, requests from collators (for example) will be eligible for authorization and will therefore fail. 

cc @Joonpark13 for vis.